### PR TITLE
Add workflow for auto publishing the repo to NPM

### PR DESCRIPTION
**Pull request by tgolen**
This PR implements a GitHub workflow that automatically publishes a new version of the library to NPM every time a PR is merged to `main`. This flow was copied from react-native-onyx.

# Tests
None. This can only be tested once it is merged. Then we will see if it works or not!

# QA
- Check the GH actions to see if it ran without any errors
- Look in https://www.npmjs.com/ to see if the package was properly published

### CODE REVIEW

File: .github/workflows/publish.yml
Comment (AndrewGable) [inline .github/workflows/publish.yml:30]: Just making sure this secret was added to the repo via Ring0?

File: .github/workflows/publish.yml
Comment (AndrewGable) [inline .github/workflows/publish.yml:69]: This secret too?

File: .github/workflows/publish.yml
Comment (luacmartins) [inline .github/workflows/publish.yml:11]: Should we add [concurrency](https://github.com/Expensify/react-native-onyx/blob/e50f029b3fe824812d14bfa25797a352afc8e60e/.github/workflows/publish.yml#L8-L10) here? I don't expect many PRs to be merged at the same time, but it might be a good check, especially in the beginning when the repo is more active.

File: .github/workflows/publish.yml
Comment (luacmartins) [inline .github/workflows/publish.yml]: I think we're moving to GH's native `concurrency` feature, instead of using turnstyle. See https://github.com/Expensify/react-native-onyx/pull/252

File: .github/workflows/publish.yml
Comment (luacmartins) [reply .github/workflows/publish.yml:30]: +1

File: .github/workflows/publish.yml
Comment (luacmartins) [reply .github/workflows/publish.yml:69]: +1

File: .github/workflows/publish.yml
Comment (tgolen) [reply .github/workflows/publish.yml:30]: It wasn't, no. Is there anything you'd be able to point me to with instructions on what Ring0 needs to do exactly?

File: .github/workflows/publish.yml
Comment (tgolen) [reply .github/workflows/publish.yml:11]: Ah, good catch. I guess I was on a slightly outdated version of the repo when I copied this. I'll update it to the latest from what's in Onyx.

File: .github/workflows/publish.yml
Comment (AndrewGable) [reply .github/workflows/publish.yml:30]: Here is an example issue: https://github.com/Expensify/Expensify/issues/288197

`LARGE_SECRET_PASSPHRASE` is in 1Password, so Ring0 will be able to find it.

**Comment by tgolen** [comment]
@AndrewGable Any idea what's wrong with the CLA workflow? https://github.com/Expensify/react-native-x-maps/actions/runs/5659066613/job/15331799891 It wasn't failing on the PR where the workflow was added.

**Comment by AndrewGable** [comment]
I think it's missing this secret too: https://github.com/Expensify/react-native-x-maps/blob/main/.github/workflows/cla.yml#L31

File: .github/workflows/publish.yml
Comment (tgolen) [reply .github/workflows/publish.yml:30]: OK, thanks. I opened this issue to have them added: https://github.com/Expensify/Expensify/issues/302806

**Review by luacmartins — APPROVED** [review]
Changes LGTM. Just need to wait on adding the secrets to the repo before we can merge this.

**Review by AndrewGable — APPROVED** [review]
Agree with @luacmartins, let's merge once secrets are added.

**Comment by tgolen** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by tgolen** [comment]
OK, tests are all passing now so I will merge and see what happens with the NPM publishing.

**Pull request merged**
